### DISCUSSION TITLE
Add support for route guards

### DISCUSSION THREAD

**Discussion by mammadataei** [TanStack/router#155] [Ideas]
Hi, I think it will be handy to support per-route and global route guards, something like how `vue-router` or nuxt middlewares work. 

```ts
const routes = [
  {
    path: '/users/:id',
    element: <User />,
    beforeEnter: (to, from, redirect) => {
      if (userCantAccess()) {
        redirect('/somewhere')
      }
    },
  },
]
```

[Navigation Guards - Vue Router](https://next.router.vuejs.org/guide/advanced/navigation-guards.html)
[Nuxt Middlewares](https://nuxtjs.org/docs/directory-structure/middleware#router-middleware)

**Comment by vincent-thomas** [comment]
I'd love to have that too! But I think it would be even nicer to have a implementation like angular guards! Its really easy to make guards there. You just return a boolean and were to navigate if false! Right now I just wrap my element in the router with a function that decides if the user is allowed to go to the specific route

**Comment by ploffredo** [comment]
Import Navigate from react-location, then use a ternary like this. 

```
const routes = [
  {
    path: '/users/:id',
    element: userCantAccess() ? <Navigate to='/somewhere' />  : <User />
  }
]
```
